cartographer/mapping_2d/sparse_pose_graph.cc

```diff
@@ -396,6 +401,23 @@ void SparsePoseGraph::WaitForAllComputations() {
   locker.Await([&notification]() { return notification; });
 }
 
+void SparsePoseGraph::FinishTrajectory(const int trajectory_id) {
+  AddWorkItem([this, trajectory_id]() REQUIRES(mutex_) {
+    CHECK_EQ(finished_trajectories_.count(trajectory_id), 0);
+    finished_trajectories_.insert(trajectory_id);
+
+    const auto& submap_data = optimization_problem_.submap_data();
```

Comment:
Can you make a pass through all submaps in `submap_data_` which match the given trajectory ID without accessing the `optimization_problem_`? Then you would not need to use the work queue at all.

Reply:
Submap states are updated within work queue thread always. See [ComputeConstraintsForScan](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping_2d/sparse_pose_graph.cc#L293), [AddSubmapFromProto](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping_2d/sparse_pose_graph.cc#L437), and [MarkSubmapAsTrimmed](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping_2d/sparse_pose_graph.cc#L670).

Reply:
Ah yes, good point. I actually posted [the same question](https://github.com/googlecartographer/cartographer/commit/7434e963697c2876add9658e1ff1af6399589215#commitcomment-24545844) to @wohe some time ago.